### PR TITLE
Add make lint and make fix-lint and make the code compliant

### PR DESCRIPTION
**Pull request by diegoceccarelli**
This commit adds two new targets to the Makefile: `make lint` and `make fix-lint`. 

`make lint` will make sure that the code is properly formatted, perform linting with `ruff`, check for typos (using `typos`, and make sure order of imports is according to the pep8 standard (using `isort`). 

Code has been fixed to pass all the checks. Next step should be set github action so that checks are performed for each PR as a prerequisite to merge into master. 

I ran tests locally and they are all successful.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -23,6 +23,23 @@ requirements: .venv  ## Install/refresh all project requirements
...
+
+.PHONY: lint
+lint: .venv  ## Check linting
+	source $(VENV_BIN)/activate
+	$(VENV_BIN)/isort --check .
+	$(VENV_BIN)/black --check .
+	$(VENV_BIN)/blackdoc .
+	$(VENV_BIN)/ruff .
+	$(VENV_BIN)/typos .
+#	$(VENV_BIN)/mypy ranx
```
Comment (diegoceccarelli) [inline Makefile:41]: This creates a lot of errors at the moment - it will take some time to make it work.

File: dev/efficiency_test.ipynb
Comment (AmenRa) [inline dev/efficiency_test.ipynb:1]: Could you check that the changes to this notebook and the others are legit and do not break them?

Thank you!

File: notebooks/1_overview.ipynb
Comment (AmenRa) [inline notebooks/1_overview.ipynb:1]: Could you check that the changes to this notebook and the others are legit and do not break them?

Thank you!

Comment (diegoceccarelli) [reply notebooks/1_overview.ipynb:1]: ack. There are five notebook modified by the PR:

1. dev/efficiency_test.ipynb
2. notebooks/1_overview.ipynb
3. notebooks/2_qrels_and_run.ipynb
4. notebooks/4_comparison_and_report.ipynb
5. notebooks/5_fusion.ipynb

I'm going to rerun all of them and make sure they behave.

Comment (AmenRa) [reply notebooks/1_overview.ipynb:1]: The efficiency test (`dev/efficiency_test.ipynb`) can take a while.
For brevity, you can just check for a single value in the for loop (5th cell).

**Comment by diegoceccarelli** [comment]
All good, I ran all the notebooks (including `efficiency_test` fully) and they still work. 

The only thing when I ran `efficiency_test` with 8 threads i got this: 

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[8], line 28
     24 results.append([query_count, "pytrec_eval", 1, "ndcg", ndcg_avg_time, 1.0])
     26 for threads in [8]:
     27     # Run metrics once to ensure they have been compiled
---> 28     evaluate(re_qrels, re_run, [f"map", f"mrr", f"ndcg"], threads=threads)
     30     x = get_ipython().run_line_magic('timeit', '-o -q evaluate(re_qrels, re_run, f"map", threads=threads)')
     31     avg_time = max(round(x.average, 3) * 1000, 1)

File ~/dev/ranx/.venv/lib/python3.9/site-packages/ranx/meta/evaluate.py:128, in evaluate(qrels, run, metrics, return_mean, threads, save_results_in_run, make_comparable)
    126     set_num_threads(1)
    127 elif threads != 0:
--> 128     set_num_threads(threads)
    130 if make_comparable and type(qrels) == Qrels and type(run) == Run:
    131     run = run.make_comparable(qrels)

File ~/dev/ranx/.venv/lib/python3.9/site-packages/numba/np/ufunc/parallel.py:608, in set_num_threads(n)
    606 if not isinstance(n, (int, np.integer)):
    607     raise TypeError("The number of threads specified must be an integer")
--> 608 snt_check(n)
    609 _set_num_threads(n)

File ~/dev/ranx/.venv/lib/python3.9/site-packages/numba/np/ufunc/parallel.py:570, in gen_snt_check.<locals>.snt_check(n)
    568 def snt_check(n):
    569     if n > NUMBA_NUM_THREADS or n < 1:
--> 570         raise ValueError(msg)

ValueError: The number of threads must be between 1 and 4
```
Removing 8 from the line: 
```
<<< for threads in [1, 2, 4, 8]:
>>> for threads in [1, 2, 4]:
``` 
Fixed. I don't think it's related to my change?

**Comment by diegoceccarelli** [comment]
Sure, I suspected it was due to my laptop configuration. Anyway not related to this PR, can we merge? 

If you want on a separate PR we could handle nicely the fact that the machine doesn't support that number of threads (log a warning and skipping).

**Comment by AmenRa** [comment]
Sorry for the delay, I supposed you had changed the number of threads in the list in the PR.
That's why I did not merge it sooner.

About the warning, I don't think it is necessary.

**Pull request merged**